pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisPartitionGroupOffset.java

```diff
@@ -0,0 +1,94 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.stream.kinesis;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.Map;
+import org.apache.pinot.spi.stream.StreamPartitionMsgOffset;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * A {@link StreamPartitionMsgOffset} implementation for the Kinesis partition group consumption
+ * A partition group consists of 1 or more shards. The KinesisCheckpoint maintains a Map of shards to the
+ * sequenceNumber.
+ * The sequenceNumber is the id (equivalent to offset in kafka) for the messages in the shard.
+ * From the Kinesis documentation:
+ * Each data record has a sequence number that is unique per partition-key within its shard.
+ * Kinesis Data Streams assigns the sequence number after you write to the stream with client.putRecords or client
+ * .putRecord.
+ * Sequence numbers for the same partition key generally increase over time.
+ * The longer the time period between write requests, the larger the sequence numbers become.
+ */
+public class KinesisPartitionGroupOffset implements StreamPartitionMsgOffset {
+  private final Map<String, String> _shardToStartSequenceMap;
+
+  public KinesisPartitionGroupOffset(Map<String, String> shardToStartSequenceMap) {
+    _shardToStartSequenceMap = shardToStartSequenceMap;
+  }
+
+  public KinesisPartitionGroupOffset(String offsetStr)
+      throws IOException {
+    _shardToStartSequenceMap = JsonUtils.stringToObject(offsetStr, new TypeReference<Map<String, String>>() {
+    });
+  }
+
+  public Map<String, String> getShardToStartSequenceMap() {
+    return _shardToStartSequenceMap;
+  }
+
+  @Override
+  public String toString() {
+    try {
+      return JsonUtils.objectToString(_shardToStartSequenceMap);
+    } catch (JsonProcessingException e) {
+      throw new IllegalStateException(
+          "Caught exception when converting KinesisCheckpoint to string: " + _shardToStartSequenceMap);
+    }
+  }
+
+  @Override
+  public KinesisPartitionGroupOffset fromString(String kinesisCheckpointStr) {
+    try {
+      return new KinesisPartitionGroupOffset(kinesisCheckpointStr);
+    } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)